646 reading output of topas simulation

Bogdan2423 opened the pull request.

Bogdan2423 commented:
For now I added "simulator type" option to simulation settings - I think it would be hard to do it another way. I also changed Runner class to get settings in constructor instead in run() method, as simulator type is required anyway for reading output.
At the moment the runner can run topas mock, as it just calls the script in provided directory, but it can't read its output.

grzanka commented on pymchelper/utils/runmc.py:
```diff
@@ -50,6 +50,7 @@ def main(args=None):
     import pymchelper
 
     parser = argparse.ArgumentParser()
+    parser.add_argument('-t', '--type', help='Simulator type: shieldhit, topas or fluka (default: shieldhit)', type=str, default='shieldhit')
```
There is another option with the same name:
![image](https://github.com/DataMedSci/pymchelper/assets/7374733/11d5890c-32a1-44f2-83a6-4daec4699f95)


Also - use an Enum for selecting simulator, the same way as `OutputDataType`

hendzeld commented on pymchelper/executor/options.py:
```diff
@@ -43,13 +49,23 @@ class SimulationSettings:
-        # discover the type of MC engine based on the type of input files/directories
-        # `self._mc_environment` is set to one of the `MCEnvironment` subclasses
-        self._mc_environment = self._discover_mc_engine(input_path)
+        # set `self._mc_environment` to the proper `MCEnvironment` subclass
+        # shieldhit is default
+        if simulator_type=='shieldhit':
+            if sys.platform == 'win32':
+                self._mc_environment = SH12AEnvironmentWindows
+            self._mc_environment = SH12AEnvironmentLinux
+        elif simulator_type == 'fluka':
+            self._mc_environment = FlukaEnvironment
+        elif simulator_type == 'topas':
+            self._mc_environment = TopasEnvironment
```
Class is publicly available, therefore:

- Seems like we have some magic strings here
- What will happen if somebody enters foobar instead `simulator_type `?
- Seems that in tests you are using enums, so maybe parameter **simulator_type** should have a SimulatorType type alias?

Additionally, in line 64 we are overriding code from line 63

grzanka commented on pymchelper/executor/options.py:
```diff
@@ -43,13 +51,24 @@ class SimulationSettings:
-    def __init__(self, input_path, simulator_exec_path=None, cmdline_opts=None):
+    def __init__(self, input_path, simulator_type=SimulatorType.shieldhit, simulator_exec_path=None, cmdline_opts=None):
+        # simulator type (shieldhit, topas or fluka)
+        self.simulator_type = simulator_type
```
```suggestion
        self.simulator_type : SimulatorType = simulator_type
```

is that correct typehint ?

Bogdan2423 replied:
yes

grzanka commented on pymchelper/executor/options.py:
```diff
@@ -43,13 +51,24 @@ class SimulationSettings:
     """
 
-    def __init__(self, input_path, simulator_exec_path=None, cmdline_opts=None):
+    def __init__(self, input_path, simulator_type=SimulatorType.shieldhit, simulator_exec_path=None, cmdline_opts=None):
```
```suggestion
    def __init__(self, input_path: str, simulator_type=SimulatorType.shieldhit, simulator_exec_path=None, cmdline_opts=None):
```

do we handle input_path as `Path` or `str` ? I suspect that `str`

Bogdan2423 replied:
str

grzanka commented on pymchelper/executor/runner.py:
```diff
@@ -96,14 +125,20 @@ def get_data(self):
+        estimators_list = []
+
+        if self.settings.simulator_type == SimulatorType.shieldhit:
+            output_files_pattern = os.path.join(self.workspace_manager.output_dir_absolute_path, "run_*", "*.bdo")
```
Use pathlib instead of os.path

grzanka commented on pymchelper/executor/runner.py:
```diff
@@ -96,14 +125,20 @@ def get_data(self):
+
+        if self.settings.simulator_type == SimulatorType.shieldhit:
+            output_files_pattern = os.path.join(self.workspace_manager.output_dir_absolute_path, "run_*", "*.bdo")
+            logging.debug("Files to merge {:s}".format(output_files_pattern))
```
```suggestion
            logging.debug("Files to merge: %s", output_files_pattern)
```

grzanka commented on pymchelper/executor/runner.py:
```diff
@@ -96,14 +125,20 @@ def get_data(self):
+            estimators_list = frompattern(output_files_pattern)
+
+        elif self.settings.simulator_type == SimulatorType.topas:
+            output_files_path = os.path.join(self.workspace_manager.output_dir_absolute_path, "run_1")
```
Use pathlib

grzanka commented on pymchelper/input_output.py:
```diff
@@ -148,15 +154,30 @@ def frompattern(pattern, error=ErrorEstimate.stderr, nan=True):
+def get_topas_estimators(output_files_path):
+    """Get Topas estimators from provided directory"""
+    estimators_list = []
+    for filename in os.listdir(output_files_path):
+        output_file_path = os.path.join(output_files_path, filename)
```
Use pathlib

grzanka commented on pymchelper/input_output.py:
```diff
@@ -148,15 +154,30 @@ def frompattern(pattern, error=ErrorEstimate.stderr, nan=True):
     return result
 
 
+def get_topas_estimators(output_files_path):
```
add type hints

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+
+    def __init__(self, filename):
+        super(TopasReader, self).__init__(filename)
+        self.directory = os.path.dirname(filename)
```
use pathlib

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+
+    def get_bins(self, dimensions, bins_data, output_data):  # skipcq: PYL-R0201
+        """Get number of bins, bin size and unit for each dimension"""
+        for dimension in dimensions:
+            pattern = f"# {dimension} in (\\d+) bin[s ] of ([\\d.]+) (\\w+)"
```
```suggestion
        pattern = f"# {dimension} in (\\d+) bin[s ] of ([\\d.]+) (\\w+)"
        for dimension in dimensions:           
```

no need to redefine pattern for every pass of the loop

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+        super(TopasReader, self).__init__(filename)
+        self.directory = os.path.dirname(filename)
+
+    def get_bins(self, dimensions, bins_data, output_data):  # skipcq: PYL-R0201
```
```suggestion
    @staticmethod
    def get_bins(dimensions, bins_data, output_data):
```

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+                return False
+        return True
+
+    def get_scorer_name(self, output_data):  # skipcq: PYL-R0201
```
```suggestion
    @staticmethod
    def get_scorer_name(output_data):
```

add type hinting as well

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+        """
+        # find the input file to extract the number of histories from it
+        num_histories = 0
+        for input_file in os.listdir(self.directory):
```
use pathlib

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+                            num_histories = int(number_str.group())
+
+        # generate estimator object for each output file
+        with open(self.filename) as output_file:
```
open the file in the reading mode.
`output_file` is a misleading name, it suggest we write something to the file.
better to name it `topas_results_file` or similar

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+                    actual_dimensions = curr_dimensions
+                    break
+
+            estimator.file_corename = os.path.basename(self.filename)[:-4]
```
use pathlib

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,172 @@
+        # find the input file to extract the number of histories from it
+        num_histories = 0
+        for input_file in os.listdir(self.directory):
+            if input_file.endswith(".txt"):
```
You can extract input file name ("parameter file") from the comment section of the output file. Also - try to check what happens if the output file is located in a different location that the input file.

Bogdan2423 replied:
if input file is not in the same directory as output file, number of histories will be set to 0

grzanka commented on tests/test_read_topas_files.py:
```diff
@@ -0,0 +1,118 @@
+
+def test_read():
+    """Test if Topas output file is read correctly"""
+    file_path = os.path.join("tests", "res", "topas", "minimal", "fluence_bp_protons_xy.csv")
```
move that to a fixture, see for example https://github.com/DataMedSci/pymchelper/blob/2a7c6a560284f42e37f0aa09ed0fe5327f83eb72/tests/test_mcpl.py#L13

grzanka commented on tests/test_read_topas_files.py:
```diff
@@ -0,0 +1,118 @@
+    assert estimator.pages[0].diff_axis1.min_val == 0
+    assert estimator.pages[0].diff_axis1.max_val == 100
+    
+    expected_data_mean = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 9.080903509193889e-11, 9.013105463113584e-11,
```
check couple of values instead of full table

grzanka commented:
@Bogdan2423 I've just merged https://github.com/DataMedSci/pymchelper/pull/658 which introduced a lot of conflicts

grzanka commented on tests/test_runner.py:
```diff
@@ -1,16 +1,25 @@
 
 
+@pytest.fixture
+def topas_path() -> Generator[Tuple[Path, Path], None, None]:
```
Either name it differently, say:
`topas_mock_and_input` or split into two fixtures.

grzanka commented on tests/test_runner.py:
```diff
@@ -1,16 +1,25 @@
+@pytest.fixture
+def topas_path() -> Generator[Tuple[Path, Path], None, None]:
+    """Return topas executable and input file paths"""
+    topas_exec_path = Path("tests") / "res" / "mocks" / "topas_minimal" / "topas"
+    topas_input_path = Path("tests") / "res" / "mocks" / "topas_minimal" / "minimal.txt"
```
extract common part of path to a variable

grzanka commented on tests/test_runner.py:
```diff
@@ -1,16 +1,25 @@
+@pytest.fixture
+def topas_path() -> Generator[Tuple[Path, Path], None, None]:
+    """Return topas executable and input file paths"""
+    topas_exec_path = Path("tests") / "res" / "mocks" / "topas_minimal" / "topas"
```
Better to use paths relative to this script, like:
```
    main_dir = Path(__file__).resolve().parent
    installation_path = main_dir / 'res' / 'shieldhit' / 'executable'
```

Bogdan2423 commented:
I made the "simulator type" argument in SimulationSettings optional and brought back guessing in case it is not provided.
I also added more tests for reader and fixed reading output files for some cases - now the topas reader should work in all cases I can think of.

hendzeld commented on pymchelper/utils/runmc.py:
```diff
@@ -102,13 +106,15 @@ def main(args=None):
     #   - location of MC simulator executable file (i.e. `shieldhit` or `rfluka`)
     #   - simulation options for the MC engine provided via -m switch (i.e. --time or -v)
     settings = SimulationSettings(input_path=parsed_args.input,
+                                  simulator_type=parsed_args.type,
```
There is no such field as `type`.

You have to name param `type` instead `simulator` (line 54).
Alternatively, you can use `dest` as it was done in line 68 ` dest='outtype'`

```suggestion
                                  simulator_type=parsed_args.simulator,
```

Bogdan2423 replied:
you're right, I think the field was called "type" earlier and I didn't change it

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,232 @@
+        self.directory = Path(filename).parent
+
+    @staticmethod
+    def get_parameter_filename(results_data: str) -> str:
```
consider migrating this to a generator approach, 
move this function outside of class
name it like `extract_parameter_filename`
apply it on a line, instead of full file content

grzanka commented on tests/test_read_topas_files.py:
```diff
@@ -0,0 +1,157 @@
+import pytest
+
+@pytest.fixture()
+def main_dir() -> Generator[Path, None, None]:
```
consider migration to conftest.py

grzanka commented on pymchelper/executor/runner.py:
```diff
@@ -45,26 +51,51 @@ def __init__(self, jobs=None, keep_workspace_after_run=False, output_directory='
+            # for that we need to modify the input file and set the number of threads to the number of jobs
+            # we set one rng seed, so one working directory and one worker will be created
+
+            modified_input_path = str(self.settings.input_path).replace(".txt", "_modified.txt")
```
Make `input_path` a Pathlib object and use `with_suffix`

Bogdan2423 replied:
I think it's not possible to change the suffix to `"_modified.txt"` using `with_suffix` method - when I use it like this: `Path(self.settings.input_path).with_suffix("_modified.txt")` it returns an error: 
`ValueError: Invalid suffix '_modified.txt'`

grzanka replied:
Thats correct. I as more concern about the fact that `.txt` may be part of directory filename (I've such cases....).
What about replacing core filename (stem?) using pathlib ?

Bogdan2423 replied:
Something like this?
`modified_input_filename = Path(self.settings.input_path).name.replace(".txt", "_modified.txt")
 modified_input_path = Path(self.settings.input_path).parent / modified_input_filename
`

grzanka commented on tests/test_runner.py:
```diff
@@ -149,3 +162,32 @@ def test_merging(example_input_cfg: dict, tmp_path: Path, monkeypatch: pytest.Mo
+
+
+@pytest.mark.smoke
+@pytest.mark.skipif(sys.platform == "darwin", reason="we don't have SHIELD-HIT12A demo binary for MacOSX")
```
```suggestion
@pytest.mark.skipif(sys.platform == "darwin", reason="we don't have TOPAS binary for MacOSX")
```

grzanka commented on tests/test_runner.py:
```diff
@@ -149,3 +162,32 @@ def test_merging(example_input_cfg: dict, tmp_path: Path, monkeypatch: pytest.Mo
+@pytest.mark.skipif(sys.platform == "darwin", reason="we don't have SHIELD-HIT12A demo binary for MacOSX")
+@pytest.mark.skipif(sys.platform == "win32", reason="simulator mocks don't work on Windows")
+def test_topas(topas_mock_path: Path, topas_input_path: Path, tmp_path: Path):
+    """Test if runner can run topas mock and read the output"""
```
```suggestion
    """Test if runner can run TOPAS mock and read the output"""
```

grzanka commented on tests/test_runner.py:
```diff
@@ -149,3 +162,32 @@ def test_merging(example_input_cfg: dict, tmp_path: Path, monkeypatch: pytest.Mo
+    assert 'fluence_bp_protons_xy' in data
+    assert 'fluence_bp_protons_xy2' in data
+    
+    shutil.rmtree(tmp_path)
```
```suggestion
```

Its bad practice to remove tmp_path. pytest always keep couple of recent directories and removes oldest using round robin algorithm.  See https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html

https://docs.pytest.org/en/7.1.x/reference/reference.html#tmp-path

![image](https://github.com/DataMedSci/pymchelper/assets/7374733/f863178c-25b0-4903-ba9b-a7161866e44a)

grzanka commented on tests/test_runner.py:
```diff
@@ -10,6 +11,17 @@
 
+@pytest.fixture
+def topas_mock_path() -> Generator[Path, None, None]:
+    """Return path to Topas mock executable"""
```
```suggestion
    """path to TOPAS mock executable"""
```

grzanka commented on tests/test_runner.py:
```diff
@@ -10,6 +11,17 @@
+    
+@pytest.fixture
+def topas_input_path() -> Generator[Path, None, None]:
+    """Return path to Topas input file"""
```
```suggestion
    """path to TOPAS input file"""
```

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,272 @@
+
+def extract_parameter_filename(header_line: str) -> Optional[str]:
+    """Get parameter filename from the output file"""
+    parameter_filename = ""
```
```suggestion
    parameter_filename = None
```

I think that returning None will more consistent with the `match.group` behaviour if not pattern is found

Bogdan2423 replied:
Yes, I forgot to remove this line - it is overwritten with None anyway a few lines later

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,272 @@
+    def read_data(self, estimator: Estimator) -> bool:
+        """
+        Read the data from the file and store them in the provided estimator object.
+        Topas reader assumes that the input file is in the same directory as the output file
```
```suggestion
        TOPAS reader assumes that the input file is in the same directory as the output file
```

grzanka commented on pymchelper/readers/topas.py:
```diff
@@ -0,0 +1,272 @@
+        or the input file is not in the same directory, the number of histories is set to 0.
+        """
+        with open(self.filename, 'r') as results_file:
+            results_lines = results_file.readlines()
```
I don't like the programming style with `break`.
What do you think of creating header lines or comment lines:

```comment_lines = [line for results_lines if line.startswith("#")]```

and
```data_lines = [line for results_lines if not line.startswith("#")]```

? Then you could adjust the code.

Bogdan2423 replied:
Yes, I think it looks better

grzanka commented on pymchelper/input_output.py:
```diff
@@ -156,15 +163,29 @@ def frompattern(pattern: str, error: ErrorEstimate = ErrorEstimate.stderr, nan:
+def get_topas_estimators(output_files_path: str) -> List[Estimator]:
+    """Get Topas estimators from provided directory"""
+    estimators_list = []
+    for result_file_path in Path(output_files_path).iterdir():
```
```suggestion
    for path in Path(output_files_path).iterdir():
```

No reason to call all files and directories a results_file_path

grzanka replied:
```suggestion
    for path in Path(output_files_path).iterdir():
```

No reason to call all files and directories a results_file_path

grzanka commented on pymchelper/executor/runner.py:
```diff
@@ -45,26 +51,52 @@ def __init__(self, jobs=None, keep_workspace_after_run=False, output_directory='
+                else:
+                    config = f"i:Ts/NumberOfThreads = {self.jobs}\n" + config
+
+            with open(modified_input_path, "w") as file:
+                file.write(config)
```
```suggestion
            modified_input_path.write_text(config)
```

would that work as well ?

Bogdan2423 replied:
Yes
